File: lib/r/mango.R
library(jsonlite)

mango_socket_path <- function() {
  p <- Sys.getenv("MANGO_INSTANCE_SIGNATURE")
  if (nzchar(p)) p else stop("MANGO_INSTANCE_SIGNATURE is not set")
}

mmsg_capture <- function(args) {
  out <- suppressWarnings(system2("mmsg", args = args, stdout = TRUE))
  rc <- attr(out, "status")
  if (!is.null(rc) && rc != 0) {
    stop("mmsg failed: ", paste(args, collapse = " "))
  }
  out
}

parse_reply <- function(text, cmd) {
  reply <- fromJSON(paste(text, collapse = "\n"))
  if ("error" %in% names(reply)) {
    stop(paste0("mango error for ", cmd, ": ", reply$error))
  }
  reply
}

mango_call <- function(cmd) {
  parse_reply(mmsg_capture(strsplit(cmd, " ", fixed = TRUE)[[1]]), cmd)
}

mango_call_argv <- function(args) {
  parse_reply(mmsg_capture(args), paste(args, collapse = " "))
}

mango_retry <- function(cmd, attempts = 50, delay = 0.1) {
  for (i in seq_len(attempts)) {
    ok <- tryCatch({ mango_call(cmd); TRUE }, error = function(e) FALSE)
    if (ok) return(invisible(TRUE))
    Sys.sleep(delay)
  }
  stop("mango not reachable after ", attempts, " attempts: ", cmd)
}

mango_get <- function(spec) mango_call(paste("get", spec))

mango_set_option <- function(key, value) {
  mango_call_argv(c("setoption", key, value))
}

mango_dispatch <- function(function_name, ...) {
  mango_call_argv(c("dispatch", paste(c(function_name, ...), collapse = ",")))
}

mango_unset_bind <- function(mode, mods, keysym, family = "bind") {
  mango_call_argv(c("unset", "bind", mode, mods, keysym, family))
}

mango_unset_rule <- function(kind, spec) {
  mango_call_argv(c("unset", kind, spec))
}

mango_watch <- function(subject, handler = NULL) {
  lines <- mmsg_capture(c("watch", subject))
  for (line in lines) {
    if (identical(trimws(line), "")) next
    event <- parse_reply(line, paste("watch", subject))
    if (is.function(handler)) handler(event) else print(event)
  }
  invisible(TRUE)
}

mango_watch_first <- function(subject) {
  lines <- mmsg_capture(c("watch", subject))
  parse_reply(lines[1], paste("watch", subject))
}

mango_version <- function() mango_get("version")$version
mango_monitors <- function() mango_get("all-monitors")$monitors
mango_clients <- function() mango_get("all-clients")$clients
mango_focused_client <- function() mango_get("focusing-client")
mango_binds <- function() mango_get("binds")$binds
mango_rules <- function() mango_get("rules")$rules
mango_option <- function(key) mango_get(paste("option", key))$value
mango_options <- function() mango_get("options")

mango_bind <- function(mode, mods, keysym, cmd) {
  mango_dispatch("setup_bind", mode, mods, keysym, cmd)
}

mango_rule <- function(match, ...) {
  mango_dispatch("setup_rule", match, ...)
}

mango_rule_for_class <- function(class_name, ...) {
  mango_rule(paste0("class=", class_name), ...)
}

mango_set_layout <- function(name) {
  mango_dispatch("setlayout", name)
}

mango_toggle_layout <- function() {
  count <- get0(".mango_layout_calls", envir = .GlobalEnv, ifnotfound = 0L)
  assign(".mango_layout_calls", count + 1L, envir = .GlobalEnv)
  mango_set_layout(if ((count + 1L) %% 2 == 1L) "tile" else "dwindle")
}

mango_borders <- function(px, color) {
  list(
    mango_set_option("borderpx", as.character(px)),
    mango_set_option("bordercolor", color)
  )
}

mango_gaps <- function(h, v) {
  list(
    mango_set_option("gappih", as.character(h)),
    mango_set_option("gappiv", as.character(v))
  )
}

mango_theme <- function(hex, border_px) {
  list(
    mango_set_option("rootcolor", hex),
    mango_set_option("bordercolor", hex),
    mango_set_option("borderpx", as.character(border_px))
  )
}

mango_watch_until <- function(subject, predicate, timeout = 5.0) {
  deadline <- Sys.time() + timeout
  repeat {
    frame <- mango_watch_first(subject)
    if (isTRUE(predicate(frame))) {
      return(frame)
    }
    if (Sys.time() > deadline) {
      stop("watch_until timeout: watch ", subject)
    }
  }
}

mango_for_each_client <- function(fn) {
  for (client in mango_clients()) {
    fn(client)
  }
  invisible(TRUE)
}

mango_focused <- function(fn) {
  client <- mango_focused_client()
  if (!is.null(client) && !is.null(client$id) && !identical(client$id, NA)) {
    fn(client)
  }
  invisible(TRUE)
}

mango_options_map <- function() {
  raw <- mango_options()
  if (!is.null(raw$options)) {
    vals <- raw$options
    if (is.data.frame(vals)) {
      return(setNames(as.list(vals$value), vals$option))
    }
    return(vals)
  }
  if (!is.null(raw$option)) {
    return(setNames(list(raw$value), raw$option))
  }
  raw
}

mango_info <- function() {
  list(
    version = mango_version(),
    monitors = mango_monitors(),
    options = mango_options()
  )
}

mango_apply_options <- function(pairs) {
  lapply(names(pairs), function(key) mango_set_option(key, as.character(pairs[[key]])))
}

File: lib/php/mango.php
<?php
class MangoConnectError extends RuntimeException {}

function mango_socket_path(): string
{
    $path = getenv("MANGO_INSTANCE_SIGNATURE");
    if ($path === false || $path === "") {
        throw new RuntimeException("MANGO_INSTANCE_SIGNATURE is not set");
    }
    return $path;
}

function mango_call(string $cmd): array
{
    $fp = @stream_socket_client("unix://" . mango_socket_path(), $errno, $errstr, 2);
    if ($fp === false) {
        throw new MangoConnectError("connect failed: $errstr ($errno)");
    }
    fwrite($fp, $cmd . "\n");
    $line = fgets($fp);
    fclose($fp);
    if ($line === false) {
        throw new RuntimeException("connection closed without reply: $cmd");
    }
    $reply = json_decode(rtrim($line), true);
    if (!is_array($reply)) {
        throw new RuntimeException("bad reply: $line");
    }
    if (isset($reply["error"])) {
        throw new RuntimeException($reply["error"]);
    }
    return $reply;
}

function mango_retry(string $cmd, int $attempts = 50, float $delay = 0.1): array
{
    for ($i = 0; $i < $attempts; $i++) {
        try {
            return mango_call($cmd);
        } catch (MangoConnectError $e) {
            if ($i + 1 < $attempts) {
                usleep((int) ($delay * 1e6));
            }
        }
    }
    throw new RuntimeException("mango not reachable after $attempts attempts: $cmd");
}

function mango_get(string ...$spec): array
{
    return mango_call("get " . implode(" ", $spec));
}

function mango_set_option(string $key, string $value): array
{
    return mango_call("setoption $key $value");
}

function mango_dispatch(string $function, string ...$args): array
{
    return mango_call("dispatch " . implode(",", array_merge([$function], $args)));
}

function mango_unset_bind(string $mode, string $mods, string $keysym, string $family = "bind"): array
{
    return mango_call("unset bind $mode $mods $keysym $family");
}

function mango_unset_rule(string $kind, string $spec): array
{
    return mango_call("unset $kind $spec");
}

function mango_watch(string $subject, ?callable $handler = null): void
{
    $fp = @stream_socket_client("unix://" . mango_socket_path(), $errno, $errstr, 2);
    if ($fp === false) {
        throw new MangoConnectError("connect failed: $errstr ($errno)");
    }
    fwrite($fp, "watch $subject\n");
    while (($line = fgets($fp)) !== false) {
        $line = rtrim($line);
        if ($line === "") {
            continue;
        }
        $event = json_decode($line, true);
        if (!is_array($event)) {
            throw new RuntimeException("bad frame: $line");
        }
        if (isset($event["error"])) {
            throw new RuntimeException($event["error"]);
        }
        if ($handler !== null && call_user_func($handler, $event) === false) {
            break;
        }
    }
    fclose($fp);
}

function mango_watch_first(string $subject): array
{
    $fp = @stream_socket_client("unix://" . mango_socket_path(), $errno, $errstr, 2);
    if ($fp === false) {
        throw new MangoConnectError("connect failed: $errstr ($errno)");
    }
    fwrite($fp, "watch $subject\n");
    $line = fgets($fp);
    fclose($fp);
    if ($line === false) {
        throw new RuntimeException("connection closed without frame: watch $subject");
    }
    $event = json_decode(rtrim($line), true);
    if (!is_array($event)) {
        throw new RuntimeException("bad frame: $line");
    }
    if (isset($event["error"])) {
        throw new RuntimeException($event["error"]);
    }
    return $event;
}

function mango_version(): ?string
{
    return mango_get("version")["version"] ?? null;
}

function mango_monitors(): array
{
    return mango_get("all-monitors")["monitors"] ?? [];
}

function mango_clients(): array
{
    return mango_get("all-clients")["clients"] ?? [];
}

function mango_focused_client(): ?array
{
    return mango_get("focusing-client");
}

function mango_binds(): array
{
    return mango_get("binds")["binds"] ?? [];
}

function mango_rules(): array
{
    return mango_get("rules")["rules"] ?? [];
}

function mango_option(string $key): ?string
{
    return mango_get("option $key")["value"] ?? null;
}

function mango_options(): array
{
    return mango_get("options");
}

function mango_bind(string $mode, string $mods, string $keysym, string $cmd): array
{
    return mango_dispatch("setup_bind", $mode, $mods, $keysym, $cmd);
}

function mango_rule(string $match, string ...$props): array
{
    return mango_dispatch("setup_rule", $match, ...$props);
}

function mango_rule_for_class(string $class, string ...$props): array
{
    return mango_rule("class=$class", ...$props);
}

function mango_set_layout(string $name): array
{
    return mango_dispatch("setlayout", $name);
}

function mango_toggle_layout(): array
{
    static $calls = 0;
    $calls++;
    return mango_set_layout($calls % 2 === 1 ? "tile" : "dwindle");
}

function mango_borders(int $px, string $color): array
{
    return [mango_set_option("borderpx", (string) $px), mango_set_option("bordercolor", $color)];
}

function mango_gaps(int $h, int $v): array
{
    return [mango_set_option("gappih", (string) $h), mango_set_option("gappiv", (string) $v)];
}

function mango_theme(string $hex, int $border_px): array
{
    return [
        mango_set_option("rootcolor", $hex),
        mango_set_option("bordercolor", $hex),
        mango_set_option("borderpx", (string) $border_px),
    ];
}

function mango_watch_until(string $subject, callable $predicate, float $timeout = 5.0): array
{
    $fp = @stream_socket_client("unix://" . mango_socket_path(), $errno, $errstr, 2);
    if ($fp === false) {
        throw new MangoConnectError("connect failed: $errstr ($errno)");
    }
    stream_set_timeout($fp, (int) $timeout);
    fwrite($fp, "watch $subject\n");
    $deadline = microtime(true) + $timeout;
    while (($line = fgets($fp)) !== false) {
        $line = rtrim($line);
        if ($line === "") {
            continue;
        }
        $frame = json_decode($line, true);
        if (!is_array($frame)) {
            throw new RuntimeException("bad frame: $line");
        }
        if (isset($frame["error"])) {
            throw new RuntimeException($frame["error"]);
        }
        if (call_user_func($predicate, $frame) === true) {
            fclose($fp);
            return $frame;
        }
        if (microtime(true) > $deadline) {
            throw new RuntimeException("watch_until timeout: watch $subject");
        }
    }
    fclose($fp);
    throw new RuntimeException("watch_until closed before match: watch $subject");
}

function mango_for_each_client(callable $fn): void
{
    foreach (mango_clients() as $client) {
        call_user_func($fn, $client);
    }
}

function mango_focused(callable $fn): void
{
    $client = mango_focused_client();
    if (isset($client["id"]) && $client["id"] !== null) {
        call_user_func($fn, $client);
    }
}

function mango_options_map(): array
{
    $raw = mango_options();
    if (isset($raw["options"]) && is_array($raw["options"])) {
        $map = [];
        foreach ($raw["options"] as $k => $v) {
            $map[$k] = isset($v["value"]) ? $v["value"] : $v;
        }
        return $map;
    }
    if (isset($raw["option"]) || isset($raw["value"])) {
        return [$raw["option"] ?? "" => $raw["value"] ?? null];
    }
    return $raw;
}

function mango_info(): array
{
    return [
        "version" => mango_version(),
        "monitors" => mango_monitors(),
        "options" => mango_options(),
    ];
}

function mango_apply_options(array $pairs): array
{
    $replies = [];
    foreach ($pairs as $key => $value) {
        $replies[] = mango_set_option((string) $key, (string) $value);
    }
    return $replies;
}